src/lazy/text/raw/reader.rs

```diff
@@ -0,0 +1,133 @@
+use crate::lazy::decoder::LazyRawReader;
+use crate::lazy::encoding::TextEncoding;
+use crate::lazy::raw_stream_item::RawStreamItem;
+use crate::lazy::text::buffer::TextBufferView;
+use crate::lazy::text::parse_result::AddContext;
+use crate::result::IonFailure;
+use crate::IonResult;
+
+/// A text Ion 1.0 reader that yields [`RawStreamItem`]s representing the top level values found
+/// in the provided input stream.
+pub struct LazyRawTextReader<'data> {
+    // The current view of the data we're reading from.
+    buffer: TextBufferView<'data>,
+    // Each time something is parsed from the buffer successfully, the caller will mark the number
+    // of bytes that may be skipped the next time the reader advances.
+    bytes_to_skip: usize,
+}
+
+impl<'data> LazyRawTextReader<'data> {
+    /// Constructs a `LazyRawTextReader` positioned at the beginning of the provided input stream.
+    pub fn new(data: &'data [u8]) -> LazyRawTextReader<'data> {
+        Self::new_with_offset(data, 0)
+    }
+
+    /// Constructs a `LazyRawTextReader` positioned at the beginning of the provided input stream.
+    /// The provided input stream is itself a slice starting `offset` bytes from the beginning
+    /// of a larger data stream. This offset is used for reporting the absolute (stream-level)
+    /// position of values encountered in `data`.
+    fn new_with_offset(data: &'data [u8], offset: usize) -> LazyRawTextReader<'data> {
+        LazyRawTextReader {
+            buffer: TextBufferView::new_with_offset(data, offset),
+            bytes_to_skip: 0,
+        }
+    }
+
+    pub fn next<'top>(&'top mut self) -> IonResult<RawStreamItem<'data, TextEncoding>>
+    where
+        'data: 'top,
+    {
+        let buffer = self.buffer;
+        if buffer.is_empty() {
+            return IonResult::incomplete("reading a top-level value", buffer.offset());
+        }
+        let (buffer_after_whitespace, _whitespace) = buffer
+            .match_optional_whitespace()
+            .with_context("skipping whitespace between top-level values", buffer)?;
+        let (remaining, matched) = buffer_after_whitespace
+            .match_top_level()
+            .with_context("reading a top-level value", buffer_after_whitespace)?;
+        // If we successfully moved to the next value, store the remaining buffer view
+        self.buffer = remaining;
+        Ok(matched)
+    }
+}
+
+impl<'data> LazyRawReader<'data, TextEncoding> for LazyRawTextReader<'data> {
+    fn new(data: &'data [u8]) -> Self {
+        LazyRawTextReader::new(data)
+    }
+
+    fn next<'a>(&'a mut self) -> IonResult<RawStreamItem<'data, TextEncoding>> {
+        self.next()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::lazy::decoder::LazyRawValue;
+    use crate::IonType;
+
+    #[test]
+    fn test_top_level() -> IonResult<()> {
```

Comment:
🗺️ This test demonstrates the total capability introduced in this PR. Namely, a `LazyRawTextReader` that can read top-level, unannotated nulls, bools, and ints in a text stream.